core/trino-main/src/main/java/io/trino/operator/DeleteAndInsertMergeProcessor.java

```diff
@@ -0,0 +1,289 @@
+                if (operation == DELETE_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
+                    addDeleteRow(pageBuilder, inputPage, position);
+                    rowIdPositions[resultPosition] = rowIdPosition;
+                    resultPosition++;
+                }
+                if (operation == INSERT_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
+                    addInsertRow(pageBuilder, mergeCaseBlocks, position);
+                    rowIdPositions[resultPosition] = 0;
+                    resultPosition++;
+                }
+            }
+            if (INTEGER.getLong(rowMatchedBlock, position) > 0) {
+                rowIdPosition++;
+            }
+        }
+        checkArgument(resultPosition == totalPositions, "resultPosition (%s) is not equal to totalPositions (%s)", resultPosition, totalPositions);
+        Page pageWithoutRowIdBlock = pageBuilder.build();
+        int pageChannels = pageWithoutRowIdBlock.getChannelCount();
+        Block[] blocksWithRowId = new Block[pageChannels + 1];
+        for (int channel = 0; channel < pageChannels; channel++) {
+            blocksWithRowId[channel] = pageWithoutRowIdBlock.getBlock(channel);
+        }
+        blocksWithRowId[pageChannels] = makeOutputRowIdBlock(rowIdBlock, totalPositions, rowIdPositions);
+
+        return new Page(blocksWithRowId);
+    }
+
+    private static Block makeOutputRowIdBlock(Block rowIdBlock, int totalPositions, int[] rowIdPositions)
+    {
+        Block underlyingBlock = getUnderlyingBlock(rowIdBlock);
+        if (rowIdBlock.allPositionsAreNull()) {
+            return getAllNullsRowIdBlock(rowIdBlock, underlyingBlock, totalPositions);
+        }
+        else {
+            if (underlyingBlock instanceof RowBlock) {
+                List<Block> newRowIdChildrenBuilder = new ArrayList<>();
+                rowIdBlock.getChildren().stream()
+                        .map(block -> block.getPositions(rowIdPositions, 0, totalPositions))
+                        .forEach(newRowIdChildrenBuilder::add);
+                return RowBlock.fromFieldBlocks(
+                        totalPositions,
+                        Optional.empty(),
+                        newRowIdChildrenBuilder.toArray(new Block[] {}));
+            }
+            else {
+                return rowIdBlock.getPositions(rowIdPositions, 0, totalPositions);
+            }
+        }
+    }
+
+    private static Block getAllNullsRowIdBlock(Block rowIdBlock, Block underlyingBlock, int positionCount)
+    {
+        boolean[] nulls = new boolean[positionCount];
+        Arrays.fill(nulls, true);
+        if (underlyingBlock instanceof RowBlock) {
+            return RowBlock.fromFieldBlocks(positionCount, Optional.of(nulls), rowIdBlock.getChildren().toArray(new Block[]{}));
+        }
+        else {
+            return ArrayBlock.fromElementBlock(positionCount, Optional.of(nulls), new int[positionCount], underlyingBlock);
+        }
```

Comment:
Shouldn't this actually depend on `rowIdType`?

also, direct use of `ArrayBlock` is not correct. Typically you would use `io.trino.spi.type.Type#createBlockBuilder(io.trino.spi.block.BlockBuilderStatus, int)` to construct a block of values for given type.

Here, however, you actually want to create a single-value NULL block (`nativeValueToBlock` may be helpful) and wrap it in a `RunLengthEncodedBlock` instead